rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,364 @@
+implementation provides default systems to manage ui input, with a
+`InputMapping` resource to customize a little bit the system's behaviors. 
+
+### Game-oriented assumptions
+
+This was not intended during implementation, but in retrospect, the `NavMenu`
+system is relatively opinionated.
+
+Indeed, we assume we are trying to build a game UI with a series of "menus" and
+that it is possible to keep track of a "trail" of buttons from a "root menu" to
+the deepest submenu we are currently browsing.
+
+The example of a fancy editor with a bunch of dockered widgets seems to break
+that assumption. However, I think we could make it so each docker has a tab,
+and we could navigate between the tabs. I think such an implementation strategy
+encourages better mouseless navigation support and it might not be a
+limitation.
+
+![Krita UI screenshot](https://user-images.githubusercontent.com/26321040/141671939-24b8a7c3-b296-4fd4-8ae0-3bbe7fe4c9a3.png)
+
+In this example, we can imagine the root `NodeBundle` having a `NavMenu`
+component, and the elements highlighted in yellow have a `Focusable` component.
+This way we create a menu to navigate between dockers. This is perfectly
+consistent with our design.
+
+### (Solved, but worth asking) Moving UI camera & 3d UI
+
+The completely decoupled implementation of the navigation system
+enables user to implement their own UI.
+It is perfectly possible to add a `Focusable` component to a 3d object
+and, for example, provide a [mouse picking] based input system
+sending the `NavRequest`,
+while using `world_to_viewport` in `MoveParam` for gamepad navigation.
+(or completely omitting gamepad support by providing a stub `MoveParam`)
+
+
+## Drawbacks and design limitations
+
+### How does this work with the spawn/despawn workflow?
+
+The current design imposes to the user that all UI nodes such a buttons and
+menus must be already loaded, even if not visible on screen.
+This is to support re-focusing the last focused element when moving between
+menus.
+The design of this RFC _hinges on_ the whole tree been loaded in ECS while
+executing navigation requests, and would work poorly if menus were spawned and
+despawned dynamically.
+From what I gather from questions asked in the `#help` discord channel,
+I think the design most people come up with naturally is to spawn and despawn
+dynamically the menus as they are traversed, which is incompatible with
+this design.
+However, in practice, in my own games, I worked around it by expanding the
+whole UI tree beyond the UI camera view and moving around the camera.
+An alternative would simply to set the `style.display` of non-active menus to
+`None` and change the style when focus enters them.
+This also fixes the already-existing 1 frame latency issue with the
+spawn/despawn design.
+
+Something that automates that could be a natural extension of the exposed API.
+It may also help end users go with the best design first.
+
+### User upheld invariants
```

Comment:
Definitely nervous about this.

Reply:
Honestly, not that bad. The two conditions that are absolutely necessary are checked and cause panics with a detailed error message in case the invariants are not held. In any case, if there were not enforced, it would not cause UB, just buggy behavior.

But maybe I can take a closer look at them and see if they are that necessary.

Reply:
IMO there are much move severe invariants already in the engine, which to a some degree is needing to be upheld by the user (GlobalTransform + Transform for hierarchy to work comes to mind). 

That in itself is not a good argument for adding more, but we are going to need a good solution for Archetype Invariants engine-wide anyway which this design would benefit from so I'm less nervous about this. 

That being said, there is plenty of value is exploring alternatives but I don't think this should be that hard of a blocker in the name of getting some velocity on core UI features.